File: WoofAndroid/konsist/src/test/kotlin/com/perrystreet/woof/konsist/designsystem/DesignSystemAtomicDependencies.kt
package com.perrystreet.woof.konsist.designsystem

import com.lemonappdev.konsist.api.declaration.KoFileDeclaration
import com.perrystreet.woof.konsist.Assertions.assertFalse
import com.perrystreet.woof.konsist.KonsistUtils
import com.perrystreet.woof.konsist.LintRuleMessage
import io.kotest.core.spec.style.BehaviorSpec

class DesignSystemAtomicDependencies : BehaviorSpec() {
    init {
        Given("A file in the atomic design system") {
            val files = KonsistUtils.atomicDesignModule.files

            Then("Primitives depend on nothing else in the design system") {
                files.inLayer("_primitives").assertFalse(message = Message) { file ->
                    file.importsLayer("_tokens", "atoms", "molecules", "organisms", "templates", "theme")
                }
            }

            Then("Tokens depend only on primitives") {
                files.inLayer("_tokens").assertFalse(message = Message) { file ->
                    file.importsLayer("atoms", "molecules", "organisms", "templates", "theme")
                }
            }

            Then("Atoms depend only on tokens and the theme") {
                files.inLayer("atoms").assertFalse(message = Message) { file ->
                    file.importsLayer("_primitives", "molecules", "organisms", "templates")
                }
            }

            Then("Molecules depend only on atoms, tokens, and the theme") {
                files.inLayer("molecules").assertFalse(message = Message) { file ->
                    file.importsLayer("_primitives", "organisms", "templates") || file.importsOtherMolecules()
                }
            }

            Then("Organisms depend only on molecules, atoms, tokens, the theme, and other organisms") {
                files.inLayer("organisms").assertFalse(message = Message) { file ->
                    file.importsLayer("_primitives", "templates")
                }
            }

            Then("Templates depend only on organisms, molecules, atoms, tokens, and the theme") {
                files.inLayer("templates").assertFalse(message = Message) { file ->
                    file.importsLayer("_primitives")
                }
            }
        }
    }

    private fun List<KoFileDeclaration>.inLayer(layer: String) = filter { it.path.contains("/atomic/$layer/") }

    private fun KoFileDeclaration.importsLayer(vararg layers: String) =
        hasImport { import -> layers.any { layer -> import.name.contains(".atomic.$layer.") } }

    private fun KoFileDeclaration.importsOtherMolecules(): Boolean {
        val ownPackage = packagee?.name.orEmpty()
        return hasImport { import ->
            import.name.contains(".atomic.molecules.") &&
                !import.name.startsWith(ownPackage) &&
                !import.name.contains(".roles.") &&
                !import.name.contains(".state.")
        }
    }

    private companion object {
        private val Message = LintRuleMessage(
            rule = "Atomic layers only depend on the layers below them: _primitives -> _tokens -> atoms -> molecules -> organisms -> templates.",
            why = """
                Strict layering is what makes the design system composable: an atom can be reused anywhere because it
                depends on nothing above it, and a molecule never hides a whole organism inside.
            """.trimIndent(),
            howToFix = "Move shared code down (to atoms or tokens) or composition up (to organisms or templates) so imports follow the dependency flow.",
            badExample = "// in atoms/text/AtomText.kt\nimport com.perrystreet.woof.designsystem.atomic.molecules.button.MolButton",
            goodExample = "// in molecules/button/MolButton.kt\nimport com.perrystreet.woof.designsystem.atomic.atoms.text.AtomText",
        )
    }
}

File: WoofIOS/SwiftPackages/HarmonizeRules/Tests/HarmonizeRulesTests/DesignSystem/DesignSystemAtomicDependencies.swift
import Foundation
import Harmonize
import HarmonizeSemantics
import Quick

final class DesignSystemAtomicDependencies: QuickSpec {
    override class func spec() {
        Given("A file in the atomic design system") {
            let files = WoofHarmonize.atomicDesignPackage.sources()

            Then("Primitives depend on nothing else in the design system") {
                files.inFolder("_Primitives").assertFalse(message: message) { file in
                    file.references(tokenPattern) || file.references(componentPattern) || file.references(themePattern)
                }
            }

            Then("Tokens depend only on primitives") {
                files.inFolder("_Tokens").assertFalse(message: message) { file in
                    file.references(componentPattern) || file.references(themePattern)
                }
            }

            Then("Atoms depend only on tokens and the theme") {
                files.inFolder("Atoms").assertFalse(message: message) { file in
                    file.references(primitivePattern) || file.references(moleculePattern) || file.references(organismPattern) || file.references(templatePattern)
                }
            }

            Then("Molecules depend only on atoms, tokens, and the theme") {
                files.inFolder("Molecules").assertFalse(message: message) { file in
                    file.references(primitivePattern) || file.references(organismPattern) || file.references(templatePattern) || file.referencesOtherMolecules()
                }
            }

            Then("Organisms depend only on molecules, atoms, tokens, the theme, and other organisms") {
                files.inFolder("Organisms").assertFalse(message: message) { file in
                    file.references(primitivePattern) || file.references(templatePattern)
                }
            }

            Then("Templates depend only on organisms, molecules, atoms, tokens, and the theme") {
                files.inFolder("Templates").assertFalse(message: message) { file in
                    file.references(primitivePattern)
                }
            }
        }
    }

    private static let primitivePattern = try! NSRegularExpression(pattern: #"\b(Color|Alpha|Spacing|Sizing|Typography)Primitives\b"#)
    private static let tokenPattern = try! NSRegularExpression(pattern: #"\b(Colors|ColorRoles|AlphaRoles|PaddingRoles|SpacingRoles|SizingRoles|Typography|TypographyRoles|MotionRoles|AspectRatioRoles)\b"#)
    private static let themePattern = try! NSRegularExpression(pattern: #"\b(ThemeImplementing|WoofTheme)\b"#)
    private static let componentPattern = try! NSRegularExpression(pattern: #"\b(Atom|Mol|Org|Template)[A-Z]\w*\b"#)
    private static let moleculePattern = try! NSRegularExpression(pattern: #"\bMol[A-Z]\w*\b"#)
    private static let organismPattern = try! NSRegularExpression(pattern: #"\bOrg[A-Z]\w*\b"#)
    private static let templatePattern = try! NSRegularExpression(pattern: #"\bTemplate[A-Z]\w*\b"#)

    private static let message = LintRuleMessage(
        rule: "Atomic layers only depend on the layers below them: _Primitives -> _Tokens -> Atoms -> Molecules -> Organisms -> Templates.",
        why: """
            Strict layering is what makes the design system composable: an atom can be reused anywhere because it
            depends on nothing above it, and a molecule never hides a whole organism inside.
            """,
        howToFix: "Move shared code down (to atoms or tokens) or composition up (to organisms or templates) so references follow the dependency flow.",
        badExample: "// in Atoms/Text/AtomText.swift\nMolButton(text: text, onTap: onTap)",
        goodExample: "// in Molecules/Button/MolButton.swift\nAtomText(text: text, textFontRole: .displayH4)"
    )
}

private extension SwiftSourceCode {
    func references(_ pattern: NSRegularExpression) -> Bool {
        source.containsMatch(of: pattern)
    }

    func referencesOtherMolecules() -> Bool {
        let ownNames = Set(structs().map(\.name) + enums().map(\.name) + classes().map(\.name))
        let pattern = try! NSRegularExpression(pattern: #"\bMol[A-Z]\w*\b"#)
        let matches = pattern.matches(in: source, range: NSRange(source.startIndex..., in: source))
        return matches.contains { match in
            guard let range = Range(match.range, in: source) else { return false }
            let name = String(source[range])
            return !ownNames.contains(name) && !name.hasSuffix("Role") && !name.hasSuffix("State")
        }
    }
}
